File: capture-server.js
#!/usr/bin/env node
// Capture server: spawn tshark and stream simplified fields via Server-Sent Events
// Falls back to `cap` (optional native module) if tshark is unavailable.
//$env:TSHARK_PATH='C:\Program Files\Wireshark\tshark.exe'
// >> npm run capture -- 8
const http = require("http");
const { spawn } = require("child_process");

const PORT = process.env.PORT || 4000;
// Allow passing interface as first arg or via IFACE env
const INTERFACE = process.argv[2] || process.env.IFACE || "Wi-Fi";

// Allow overriding tshark binary via env or --tshark-path <path>
function resolveTsharkBin() {
  if (process.env.TSHARK_PATH) return process.env.TSHARK_PATH;
  const idx = process.argv.indexOf("--tshark-path");
  if (idx !== -1 && process.argv.length > idx + 1) return process.argv[idx + 1];
  return "tshark";
}
const TSHARK_BIN = resolveTsharkBin();

function createTsharkArgs(iface) {
  // Request many common fields so we can build a Wireshark-like details object
  return [
    "-i",
    iface,
    "-l",
    "-T",
    "fields",
    // frame
    "-e",
    "frame.number",
    "-e",
    "frame.time_relative",
    "-e",
    "frame.len",
    "-e",
    "frame.cap_len",
    // ethernet
    "-e",
    "eth.src",
    "-e",
    "eth.dst",
    // ip
    "-e",
    "ip.src",
    "-e",
    "ip.dst",
    "-e",
    "ip.proto",
    // tcp/udp
    "-e",
    "tcp.srcport",
    "-e",
    "tcp.dstport",
    "-e",
    "tcp.flags",
    "-e",
    "udp.srcport",
    "-e",
    "udp.dstport",
    // application-layer
    "-e",
    "http.request.method",
    "-e",
    "http.request.uri",
    "-e",
    "dns.qry.name",
    // generic
    "-e",
    "frame.protocols",
    "-e",
    "_ws.col.Info",
    // raw hex payload when available
    "-e",
    "data.data",
    "-E",
    "header=y",
    "-E",
    "separator=\t",
  ];
}

const server = http.createServer((req, res) => {
  if (req.url !== "/events") {
    res.writeHead(404, { "Content-Type": "text/plain" });
    return res.end("Not found");
  }

  res.writeHead(200, {
    "Content-Type": "text/event-stream",
    "Cache-Control": "no-cache",
    Connection: "keep-alive",
    "Access-Control-Allow-Origin": "*",
  });
  res.write("\n");

  console.log(`Client connected, starting capture on interface ${INTERFACE}`);

  const args = createTsharkArgs(INTERFACE);
  let tshark;
  try {
    tshark = spawn(TSHARK_BIN, args);
  } catch (e) {
    tshark = null;
  }

  if (tshark) {
    let firstLine = true;
    let headers = [];
    tshark.stdout.setEncoding("utf8");
    tshark.stdout.on("data", (chunk) => {
      const lines = chunk.split(/\r?\n/);
      for (let line of lines) {
        if (!line) continue;
        // header row
        if (firstLine) {
          headers = line.trim().split("\t");
          firstLine = false;
          continue;
        }
        const parts = line.trim().split("\t");
        const obj = {};
        for (let i = 0; i < headers.length; i++) {
          obj[headers[i]] = parts[i] || "";
        }

        // Build details object similar to Wireshark tree
        const details = {};
        // Frame
        details.frame = {
          title: `Frame ${obj["frame.number"] || ""}: ${
            obj["frame.len"] || ""
          } bytes on wire`,
          fields: [
            `Arrival Time: ${obj["frame.time_relative"] || ""}`,
            `Frame Length: ${obj["frame.len"] || ""} bytes`,
            `Capture Length: ${obj["frame.cap_len"] || ""} bytes`,
            `Frame Number: ${obj["frame.number"] || ""}`,
            `Protocols in frame: ${obj["frame.protocols"] || ""}`,
          ],
        };

        // Ethernet
        if (obj["eth.src"] || obj["eth.dst"]) {
          details.ethernet = {
            title: "Ethernet II",
            fields: [
              `Destination: ${obj["eth.dst"] || ""}`,
              `Source: ${obj["eth.src"] || ""}`,
            ],
          };
        }

        // IP
        if (obj["ip.src"] || obj["ip.dst"]) {
          details.ipv4 = {
            title: "Internet Protocol",
            fields: [
              `Source Address: ${obj["ip.src"] || ""}`,
              `Destination Address: ${obj["ip.dst"] || ""}`,
              `Protocol: ${obj["ip.proto"] || ""}`,
            ],
          };
        }

        // TCP
        if (obj["tcp.srcport"] || obj["tcp.dstport"]) {
          details.tcp = {
            title: "Transmission Control Protocol",
            fields: [
              `Source Port: ${obj["tcp.srcport"] || ""}`,
              `Destination Port: ${obj["tcp.dstport"] || ""}`,
              `Flags: ${obj["tcp.flags"] || ""}`,
            ],
          };
        }

        // UDP
        if (obj["udp.srcport"] || obj["udp.dstport"]) {
          details.udp = {
            title: "User Datagram Protocol",
            fields: [
              `Source Port: ${obj["udp.srcport"] || ""}`,
              `Destination Port: ${obj["udp.dstport"] || ""}`,
            ],
          };
        }

        // Application layer: HTTP/DNS
        if (obj["http.request.method"] || obj["http.request.uri"]) {
          details.application = {
            title: "HTTP",
            fields: [
              `Method: ${obj["http.request.method"] || ""}`,
              `URI: ${obj["http.request.uri"] || ""}`,
            ],
          };
        }
        if (obj["dns.qry.name"]) {
          details.dns = {
            title: "DNS",
            fields: [`Query: ${obj["dns.qry.name"]}`],
          };
        }

        const pkt = {
          id: obj["frame.number"] ? Number(obj["frame.number"]) : null,
          time: obj["frame.time_relative"] || "0.000000",
          timestamp: Date.now(),
          source: obj["ip.src"] || "",
          destination: obj["ip.dst"] || "",
          protocol: obj["frame.protocols"] || "",
          length: obj["frame.len"] ? Number(obj["frame.len"]) : 0,
          info: obj["_ws.col.Info"] || "",
          raw: obj["data.data"] || "",
          details,
        };
        res.write("data: " + JSON.stringify(pkt) + "\n\n");
      }
    });

    tshark.stderr.on("data", (d) =>
      console.error("tshark stderr:", d.toString())
    );

    tshark.on("error", (err) => {
      console.error("Failed to start tshark:", err.message);
      res.write(
        "data: " +
          JSON.stringify({
            error:
              "tshark failed: " +
              err.message +
              ". Provide a full path via TSHARK_PATH or --tshark-path",
          }) +
          "\n\n"
      );
      startCapFallback(res, INTERFACE);
    });
  } else {
    startCapFallback(res, INTERFACE);
  }

  req.on("close", () => {
    console.log("Client disconnected, stopping capture");
    try {
      if (tshark) tshark.kill();
    } catch (e) {}
    try {
      if (capInstance && typeof capInstance.close === "function")
        capInstance.close();
    } catch (e) {}
    res.end();
  });
});

// Try binding to requested port, otherwise increment
let startPort = Number(PORT) || 4000;
const MAX_ATTEMPTS = 50;
function tryListen(port, attempt = 0) {
  server.once("error", (err) => {
    if (err && err.code === "EADDRINUSE" && attempt < MAX_ATTEMPTS) {
      console.warn(`Port ${port} in use, trying ${port + 1}...`);
      tryListen(port + 1, attempt + 1);
    } else {
      console.error("Failed to bind port:", err);
      process.exit(1);
    }
  });
  server.once("listening", () => {
    const addr = server.address();
    const p = typeof addr === "object" && addr ? addr.port : port;
    console.log(`Capture server listening on http://localhost:${p}/events`);
    console.log(
      "Tshark args example:",
      TSHARK_BIN,
      createTsharkArgs(INTERFACE).join(" ")
    );
    if (TSHARK_BIN !== "tshark") {
      console.log(`Using tshark binary: ${TSHARK_BIN}`);
    }
  });
  server.listen(port);
}
tryListen(startPort);

// cap fallback
let capInstance = null;
function startCapFallback(res, ifaceName) {
  let Cap;
  try {
    Cap = require("cap").Cap;
  } catch (e) {
    console.error("cap missing:", e.message);
    res.write(
      "data:" +
        JSON.stringify({ error: "cap module missing: " + e.message }) +
        "\n\n"
    );
    return;
  }
  const { decoders } = require("cap");
  const PROTOCOL = decoders.PROTOCOL;

  const devices = Cap.deviceList ? Cap.deviceList() : [];
  if (!devices.length) {
    res.write(
      "data:" + JSON.stringify({ error: "No capture devices found" }) + "\n\n"
    );
    return;
  }
  let device = devices.find(
    (d) =>
      d.name.includes(ifaceName) ||
      (d.description && d.description.includes(ifaceName))
  );
  if (!device) device = devices[0];

  try {
    const c = new Cap();
    capInstance = c;
    const filter = "";
    const bufSize = 10 * 1024 * 1024;
    const buffer = Buffer.alloc(65535);
    const linkType = c.open(device.name, filter, bufSize, buffer);
    c.setMinBytes && c.setMinBytes(0);
    console.log("cap fallback capturing on", device.name, linkType);
    c.on("packet", function (nbytes, trunc) {
      try {
        let src = "",
          dst = "",
          proto = "";
        try {
          if (linkType === "ETHERNET") {
            const ret = decoders.Ethernet(buffer);
            if (ret.info.type === PROTOCOL.ETHERNET.IPV4) {
              const ip = decoders.IPV4(buffer, ret.offset);
              src = ip.info.srcaddr;
              dst = ip.info.dstaddr;
              proto = ip.info.protocol;
            }
          }
        } catch (e) {}
        const pkt = {
          id: null,
          time: (Date.now() / 1000).toFixed(6),
          source: src,
          destination: dst,
          protocol: String(proto),
          length: nbytes,
          info: trunc ? "truncated" : "raw",
          raw: buffer.slice(0, nbytes).toString("hex"),
        };
        res.write("data: " + JSON.stringify(pkt) + "\n\n");
      } catch (err) {
        console.error("cap packet error", err);
      }
    });
  } catch (err) {
    console.error("cap open failed", err);
    res.write(
      "data:" +
        JSON.stringify({ error: "cap open failed: " + (err.message || err) }) +
        "\n\n"
    );
  }
}

File: capture-server.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/google/gopacket"
	"github.com/google/gopacket/layers"
	"github.com/google/gopacket/pcap"
)

type PacketEvent struct {
	Time        string `json:"time"`
	Source      string `json:"source"`
	Destination string `json:"destination"`
	Protocol    string `json:"protocol"`
	Length      int    `json:"length"`
	Info        string `json:"info"`
}

var (
	iface   = flag.String("i", "", "Interface name to capture (use --list to list)")
	port    = flag.Int("port", 4000, "HTTP port for SSE")
	listIFs = flag.Bool("list", false, "List available pcap devices")
)

func listDevices() {
	devs, err := pcap.FindAllDevs()
	if err != nil {
		log.Fatalf("Failed to list devices: %v", err)
	}
	for _, d := range devs {
		fmt.Printf("Name: %s\n", d.Name)
		if d.Description != "" {
			fmt.Printf("  Desc: %s\n", d.Description)
		}
		for _, a := range d.Addresses {
			fmt.Printf("  Addr: %s\n", a.IP)
		}
		fmt.Println()
	}
}

func main() {
	flag.Parse()
	if *listIFs {
		listDevices()
		return
	}
	if *iface == "" {
		log.Fatalf("Interface required. Use -i <iface> or --list to enumerate interfaces.")
	}

	// verify interface exists
	ifaces, err := pcap.FindAllDevs()
	if err != nil {
		log.Fatalf("pcap.FindAllDevs error: %v", err)
	}
	found := false
	for _, d := range ifaces {
		if d.Name == *iface || strings.Contains(d.Name, *iface) || strings.Contains(d.Description, *iface) {
			found = true
			break
		}
	}
	if !found {
		log.Printf("Warning: interface '%s' not found in devices. Continuing and letting pcap try to open it.", *iface)
	}

	http.HandleFunc("/events", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/event-stream")
		w.Header().Set("Cache-Control", "no-cache")
		w.Header().Set("Connection", "keep-alive")
		w.Header().Set("Access-Control-Allow-Origin", "*")

		// open pcap handle
		handle, err := pcap.OpenLive(*iface, 65535, true, pcap.BlockForever)
		if err != nil {
			log.Printf("pcap.OpenLive error: %v", err)
			msg := map[string]string{"error": fmt.Sprintf("pcap open failed: %v", err)}
			b, _ := json.Marshal(msg)
			fmt.Fprintf(w, "data: %s\n\n", b)
			return
		}
		defer handle.Close()

		packetSource := gopacket.NewPacketSource(handle, handle.LinkType())
		ch := packetSource.Packets()

		notify := w.(http.CloseNotifier).CloseNotify()
		for {
			select {
			case <-notify:
				log.Println("Client disconnected from SSE")
				return
			case p, ok := <-ch:
				if !ok {
					log.Println("packet channel closed")
					return
				}
				pe := PacketEvent{Time: time.Now().Format(time.RFC3339Nano), Length: len(p.Data())}
				if net := p.NetworkLayer(); net != nil {
					src, dst := net.NetworkFlow().Endpoints()
					pe.Source = src.String()
					pe.Destination = dst.String()
				}
				if tr := p.TransportLayer(); tr != nil {
					pe.Protocol = tr.LayerType().String()
				} else if app := p.ApplicationLayer(); app != nil {
					pe.Protocol = app.LayerType().String()
				} else if p.Layer(layers.LayerTypeIPv4) != nil {
					pe.Protocol = "IPv4"
				}
				// simple info for TCP/UDP ports
				if tcp := p.Layer(layers.LayerTypeTCP); tcp != nil {
					t := tcp.(*layers.TCP)
					pe.Info = fmt.Sprintf("%d → %d", t.SrcPort, t.DstPort)
				} else if udp := p.Layer(layers.LayerTypeUDP); udp != nil {
					u := udp.(*layers.UDP)
					pe.Info = fmt.Sprintf("%d → %d", u.SrcPort, u.DstPort)
				}

				b, _ := json.Marshal(pe)
				fmt.Fprintf(w, "data: %s\n\n", b)
				// flush
				if f, ok := w.(http.Flusher); ok {
					f.Flush()
				}
			}
		}
	})

	// graceful shutdown on Ctrl+C
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-c
		log.Println("Shutting down capture server")
		os.Exit(0)
	}()

	addr := fmt.Sprintf(":%d", *port)
	log.Printf("Starting Go capture server on %s (iface=%s)", addr, *iface)
	log.Fatal(http.ListenAndServe(addr, nil))
}
